OutSystems/outsystems-datagrid-reactive
# RGRIDT-683: Implement Cell Edit Validation

This PR is for RGRIDT-683: Implement Cell Edit Validation

### What was happening
* There was no way to add actions that would be triggered when the event of cell edition ends. 
* There was no validation on the cells by column type

### What was done
* Added a new record into ColumnEvent static entity - OnCellValueChange
![new1](https://user-images.githubusercontent.com/6432232/110643780-5d9f2180-81ac-11eb-94c0-6e0c00b576f5.PNG)
* Added a new event and a callback that is subscribed to the column events in all columns beside Action Column.
![new2](https://user-images.githubusercontent.com/6432232/110644426-fafa5580-81ac-11eb-8943-863e248d476e.PNG)
The input parameters of this callback are → 
**GridWidgetId** → ID of Grid block that was edited. 
**RowNumber** → Index of the row that was edited.
**ColumnWidgetId** → ID of the column block that was edited. 
**OldValue** → Value of the cell before the edition.
**NewValue** → New value of the cell after the edition.
* Regarding the Javascript, we have multiple ways to trigger the event mentioned above. You can see them in the code, and they are → cellEditEnded, pastedCell, undoingAction, redoingAction. I have prepared the code for another event - addedRows, which can be used in the future.
* Added a new method to the API → SetValidationStatus which can be found inside API_Cells. 
![new4](https://user-images.githubusercontent.com/6432232/110646568-e919b200-81ae-11eb-8a63-eac1e3374824.PNG)
This method can be added to the OS flow and used to validate any cell that has been edited. Associating the method to the event OnCellValueChange for instance. It will then invoke all the methods responsible for adding the metadata (validationMark) to the row which can then be verified inside a formatItem handler to add or remove the CSS class responsible for adding a red flag to the cell as well as to the row it belongs to.
* When the mouse hovers the cell with the error I added a tooltip with a different class to show the message error in a red tooltip with the text in white. If the developer never adds a message error to the SetCellValidation, the default text should be "Invalid [Column Name]". For this one, I understand that if we wanted to have the column name it had to be done in our JS as well. So you can see two modifications on the errorMessage. The first one to validate that if the text is empty then we want to show the default message (mentioned before) and the second one to replace all end of lines with a <br> which can then be interpreted by wijmo showTooltip method.
----
➕**Plus:** 

* ✅ Regarding the multiple validations, the developer would be able to add all the validations he needed for each column. From what was stated in the last section, it would be as simple as adding more conditions to the logic of the action. To show all the errors at once. I believe that this can be accomplished by setting the text that will appear in the tooltip of the cell, this is the error message from SetValidationStatus
![image](https://user-images.githubusercontent.com/6432232/110923856-43368680-8319-11eb-96fe-454c81236b31.png)

----

### Test Steps

1.Press Delete Key on the first cell from column Name.
**Expected**: The cell should have a red border and mark. On mouseover, a tooltip should show the message "The value should not be empty. The value's length should not be smaller than 5."

1.Edit the first cell from column Price with the value 5000.
**Expected**: The cell should have a dirty flag. On mouseover, nothing should happen. On the top of the page, there should be a feedback message showing both old and new values.

1.Press Delete Key on the first cell from column Description.
**Expected**: The cell should have a red border and mark. On mouseover, a tooltip should show the message "Invalid Description" (this is the default error message when there is no custom message error)

1.Edit the first cell from column Stock with the value 200.
**Expected**: The cell should have a red border and mark. On mouseover, a tooltip should show the message "The value should be smaller than 100."

1.Edit a cell where the value is True from column Is Favourite and make it be False.
2.The cell should now have a dirty flag. **Edit the cell again** and make its value be True.
**Expected**: The cell should have a red border and mark. On mouseover, a tooltip should show the message "The value should be False".

1.Edit the first cell from column Date with any value different from today.
**Expected**: The cell should have a red border and mark. On mouseover, a tooltip should show the message "The value should have today's date".

1.Edit the first cell from column Product category with value "Laptops".
**Expected**: The cell should have a red border and mark. On mouseover, a tooltip should show the message "The value should be Headphones".

1.Edit the first cell from column DateTime with today's date.
**Expected**: The cell should have a red border and mark. On mouseover, a tooltip should show the message "The day should be different from today".

1.On a cell from Column Name (Column with 2 validations - cell can't be empty and cell should not have more than 5 chars) paste another value with less than 5 chars.
**Expected**: The cell should have a red border and mark. On mouseover, a tooltip should show the message "The value's length should not be smaller than 5."

1.On a cell from Column Name (Column with 2 validations - cell can't be empty and cell should not have more than 5 chars) paste another value with less than 5 chars.
2.Change a cell from the Text Column to "This will pass".
3.Change the same cell from the Text Column to "fail".
4.Undo actions with Ctrl + Z
**Expected**: The cell should have a dirty mark and the row header as well. No invalid marks should appear.

1.On a cell from Column Name (Column with 2 validations - cell can't be empty and cell should not have more than 5 chars) paste another value with less than 5 chars.
2.Change a cell from the Text Column to "This will pass".
3.Change the same cell from the Text Column to "fail".
4.Undo actions with Ctrl + Z
5.Redo actions with Ctrl + Y
**Expected**: The cell should have a red border and mark. On mouseover, a tooltip should show the message "The value's length should not be smaller than 5."

### Screenshots
![edit](https://user-images.githubusercontent.com/6432232/110651645-8840a880-81b3-11eb-8775-76d5acbcd16d.gif)


### Checklist
* [x] requires changes in OutSystems (if so, provide a module with changes) - **RGRIDT-683**
* [] requires new sample page in OutSystems (if so, provide a module with changes)
